src/main/java/com/thealgorithms/strings/AhoCorasick.java

```diff
@@ -0,0 +1,181 @@
+    public void buildTrie(String[] patterns) {
+        root = new Node(); // Initialize the root of the trie
+        res = new ArrayList<>(patterns.length); // Initialize the result data structure
+        this.patterns = patterns;
+        // Loop through each input pattern
+        for (int i = 0; i < patterns.length; i++) {
+            res.add(new ArrayList<Integer>()); // Initialize a list to store positions of the current pattern
+
+            Node curr = root; // Start at the root of the trie for each pattern
+
+            // Loop through each character in the current pattern
+            for (int j = 0; j < patterns[i].length(); j++) {
+                char c = patterns[i].charAt(j); // Get the current character
+
+                // Check if the current node has a child for the current character
+                if (curr.child.containsKey(c)) {
+                    curr = curr.child.get(c); // Update the current node to the child node
+                } else {
+                    // If no child node exists, create a new one and add it to the current node's children
+                    Node nn = new Node();
+                    curr.child.put(c, nn);
+                    curr = nn; // Update the current node to the new child node
+                }
+            }
+            curr.pattern_ind = i; // Store the index of the pattern in the current leaf node
+        }
+    }
+
+    // Builds the suffix links and output links in the Aho-Corasick trie
+    public void buildSuffixAndOutputLinks() {
+        root.suffix_link = root; // Initialize the suffix link of the root to itself
+        Queue<Node> q = new LinkedList<>(); // Initialize a queue for BFS traversal
+
+        // Initialize suffix links for child nodes of the root
+        for (char rc : root.child.keySet()) {
+            q.add(root.child.get(rc)); // Add child node to the queue
+            root.child.get(rc).suffix_link = root; // Set suffix link to the root
+        }
+
+        while (!q.isEmpty()) {
+            Node currentState = q.poll(); // Get the current node for processing
+
+            // Iterate through child nodes of the current node
+            for (char cc : currentState.child.keySet()) {
+                Node currentChild = currentState.child.get(cc); // Get the child node
+                Node parentSuffix = currentState.suffix_link; // Get the parent's suffix link
+
+                // Calculate the suffix link for the child based on parent's suffix link
+                while (!parentSuffix.child.containsKey(cc) && parentSuffix != root) {
+                    parentSuffix = parentSuffix.suffix_link;
+                }
+
+                // Set the calculated suffix link or default to root
+                if (parentSuffix.child.containsKey(cc)) {
+                    currentChild.suffix_link = parentSuffix.child.get(cc);
+                } else {
+                    currentChild.suffix_link = root;

```

Comment:
This is not covered by any of the tests.

Reply:
Got it. Will add a test regarding this.

Reply:
Added test-case, regarding this with few more test-cases.
It is also done now.